File: sdk-java/src/main/java/io/superdurable/dex/Wait.java
/*
 * Legacy Materials in this file remain under their original licenses.
 * See LEGACY_NOTICES.md.
 */

/*
 * Modifications Copyright (c) 2026 Super Durable, Inc.
 *
 * Modifications after the Legacy Cutoff are licensed under the
 * Super Durable Source License 1.0.
 * Legacy Materials remain under their original licenses.
 * See LICENSE and LEGACY_NOTICES.md.
 */

package io.superdurable.dex;

import io.superdurable.dex.exceptions.InvalidStepResultException;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

/**
 * Describes when Dex may invoke a Step's execute method.
 *
 * <p>Return a {@code Wait} from {@link Step#waitFor}. The value is declarative and immutable: Dex
 * persists and evaluates its Timer, Channel, and SubFlow conditions without blocking the worker
 * thread that produced it. Use {@link #until} for a single condition, {@link #allOf} or
 * {@link #anyOf} for a flat group, and {@link #anyCombinationOf} for alternatives containing
 * multiple conditions.
 *
 * <pre>{@code
 * public Wait waitFor(Context context, Order input) {
 *     return Wait.anyOf(
 *             paymentReceived.forOne("payment"),
 *             Timer.byDuration(Duration.ofHours(1), "timeout"));
 * }
 * }</pre>
 */
public final class Wait {
    enum Kind {
        SKIP_IMMEDIATELY,
        ALL_OF,
        ANY_OF,
        ANY_COMBINATION_OF
    }

    private final Kind kind;
    private final List<Condition> conditions;
    private final List<ConditionCombination> combinations;

    private Wait(
            final Kind kind,
            final List<Condition> conditions,
            final List<ConditionCombination> combinations) {
        this.kind = kind;
        this.conditions = Collections.unmodifiableList(conditions);
        this.combinations = Collections.unmodifiableList(combinations);
    }

    /**
     * Skips condition evaluation and makes the Step immediately eligible to execute.
     *
     * @return an immediate wait definition
     */
    public static Wait skipImmediately() {
        return ofConditions(Kind.SKIP_IMMEDIATELY);
    }

    /**
     * Waits for one condition to be satisfied.
     *
     * @param condition the single Timer, Channel, or SubFlow condition
     * @return a wait equivalent to {@code allOf(condition)}
     */
    public static Wait until(final Condition condition) {
        return allOf(condition);
    }

    /**
     * Waits until every supplied condition is satisfied.
     *
     * @param conditions the conditions that must all be satisfied
     * @return an all-of wait definition
     */
    public static Wait allOf(final Condition... conditions) {
        return ofConditions(Kind.ALL_OF, conditions);
    }

    /**
     * Waits until any supplied condition is satisfied.
     *
     * <p>Channel consumption is not greedy across alternatives. Dex consumes messages only from
     * the selected Channel condition; other ready Channel conditions consume nothing.
     *
     * @param conditions the alternative conditions
     * @return an any-of wait definition
     */
    public static Wait anyOf(final Condition... conditions) {
        return ofConditions(Kind.ANY_OF, conditions);
    }

    /**
     * Waits until every condition in any one supplied combination is satisfied.
     *
     * <p>Channel consumption is not greedy across combinations. Dex consumes messages only from
     * Channel conditions in the selected combination. Conditions belonging only to other ready
     * combinations consume nothing.
     *
     * <p>Every Condition must have a non-empty user-provided ID. Reusing the same Condition object
     * across combinations is supported; distinct Conditions must not share an ID.
     *
     * @param combinations alternative all-of condition groups
     * @return an any-combination wait definition
     * @throws InvalidStepResultException when a Condition ID is missing, empty, or duplicated
     */
    public static Wait anyCombinationOf(final ConditionCombination... combinations) {
        return new Wait(
                Kind.ANY_COMBINATION_OF,
                Collections.<Condition>emptyList(),
                Arrays.asList(combinations.clone()));
    }

    private static Wait ofConditions(final Kind kind, final Condition... conditions) {
        return new Wait(
                kind,
                Arrays.asList(conditions.clone()),
                Collections.<ConditionCombination>emptyList());
    }

    Kind getKind() {
        return kind;
    }

    List<Condition> getConditions() {
        return conditions;
    }

    List<ConditionCombination> getCombinations() {
        return combinations;
    }
}

File: sdk-python/dex/wait.py
# Legacy Materials in this file remain under their original licenses.
# See LEGACY_NOTICES.md.

# Modifications Copyright (c) 2026 Super Durable, Inc.
#
# Modifications after the Legacy Cutoff are licensed under the
# Super Durable Source License 1.0.
# Legacy Materials remain under their original licenses.
# See LICENSE and LEGACY_NOTICES.md.

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum

from dex.condition import Condition, ConditionCombination


class WaitKind(Enum):
    SKIP_IMMEDIATELY = "skip_immediately"
    ALL_OF = "all_of"
    ANY_OF = "any_of"
    ANY_COMBINATION_OF = "any_combination_of"


@dataclass(frozen=True)
class Wait:
    """Describe the durable conditions evaluated before a Step executes.

    A Step returns a Wait from ``wait_for``. Dex persists the condition tree and
    invokes ``execute`` only when the selected readiness rule succeeds.

    Attributes:
        kind: The combination rule applied by Dex.
        conditions: Direct Conditions for all-of or any-of waits.
        combinations: Alternative all-of groups for any-combination waits.

    Examples:
        >>> wait = Wait.any_of(
        ...     replies.for_one(),
        ...     Timer.by_duration(timedelta(hours=1)),
        ... )
    """

    kind: WaitKind
    conditions: tuple[Condition, ...] = ()
    combinations: tuple[ConditionCombination, ...] = ()

    @staticmethod
    def skip_immediately() -> Wait:
        """Create a Wait that proceeds to ``execute`` immediately.

        Returns:
            A Wait with no Conditions.
        """
        return Wait(WaitKind.SKIP_IMMEDIATELY)

    @staticmethod
    def until(condition: Condition) -> Wait:
        """Create a Wait for one Condition.

        The Condition does not need a condition ID.

        Args:
            condition: The only readiness condition.

        Returns:
            An all-of Wait with ``condition``.
        """
        return Wait.all_of(condition)

    @staticmethod
    def all_of(*conditions: Condition) -> Wait:
        """Create a Wait that requires every Condition.

        This combinator does not require condition IDs. Every Condition in
        :meth:`any_combination_of` does require one.

        Args:
            *conditions: Conditions evaluated as one all-of group.

        Returns:
            An all-of Wait with Conditions in argument order.
        """
        return Wait(WaitKind.ALL_OF, conditions)

    @staticmethod
    def any_of(*conditions: Condition) -> Wait:
        """Create a Wait that continues when any Condition is ready.

        Conditions do not need condition IDs. Read Channel results from the
        Channel definition and inspect Timer outcomes through ``Context``.

        Channel consumption is not greedy across alternatives. Dex consumes
        messages only from the selected Channel Condition; other ready Channel
        Conditions consume nothing.

        Args:
            *conditions: Alternative readiness Conditions.

        Returns:
            An any-of Wait with Conditions in argument order.
        """
        return Wait(WaitKind.ANY_OF, conditions)

    @staticmethod
    def any_combination_of(*combinations: ConditionCombination) -> Wait:
        """Create a Wait that accepts any complete Condition combination.

        Channel consumption is not greedy across combinations. Dex consumes
        messages only from Channel Conditions in the selected combination;
        Conditions belonging only to other ready combinations consume nothing.

        Args:
            *combinations: Alternative all-of groups created with
                :meth:`ConditionCombination.of`.

        Returns:
            A Wait containing the alternative groups.
        """
        return Wait(
            WaitKind.ANY_COMBINATION_OF,
            combinations=combinations,
        )
